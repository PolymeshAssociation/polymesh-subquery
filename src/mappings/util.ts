import { decodeAddress } from '@polkadot/keyring';
import { Codec } from '@polkadot/types/types';
import { hexStripPrefix, u8aToHex, u8aToString } from '@polkadot/util';
import { SubstrateEvent, SubstrateExtrinsic } from '@subql/types';
import { Portfolio } from 'polymesh-subql/types/models/Portfolio';
import {
  AssetDocument,
  Compliance,
  Distribution,
  FoundType,
  SecurityIdentifier,
  TransferComplianceExemption,
  TransferManager,
  TransferRestrictionTypeEnum,
} from '../types';
import { Attributes } from './entities/common';

/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str[0].toLowerCase() + str.slice(1).replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`);

export const snakeToCamelCase = (value: string): string =>
  value
    .toLowerCase()
    .replace(/([-_][a-z])/g, group => group.toUpperCase().replace('-', '').replace('_', ''));

export const capitalizeFirstLetter = (str: string): string => str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s?.replace(/\0/g, '') || '';

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (text: string, exitOnFirst = false): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case '(':
      case '{':
      case '<': {
        nestedLevel++;
        break;
      }
      case ')':
      case '}':
      case '>': {
        nestedLevel--;
        break;
      }
      case ',': {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(`No top level comma found in ${text}, it probably isn't a map`);
  }
  return commas;
};

export const getOrDefault = <K, V>(map: Map<K, V>, key: K, getDefault: () => V): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};

export const serializeAccount = (item: Codec): string | null => {
  const s = item.toString();

  if (s.trim().length === 0) {
    return null;
  }
  return u8aToHex(decodeAddress(item.toString(), false, item.registry.chainSS58));
};

export const getFirstKeyFromJson = (item: Codec): string => {
  return Object.keys(item.toJSON())[0];
};

export const getFirstValueFromJson = (item: Codec): string => {
  return item.toJSON()[getFirstKeyFromJson(item)];
};

export const getTextValue = (item: Codec): string => {
  return item?.toString().trim().length === 0 ? null : item.toString().trim();
};

export const getNumberValue = (item: Codec): number => {
  return Number(getTextValue(item));
};

export const getDateValue = (item: Codec): Date => {
  return item.toString().trim().length === 0 ? null : new Date(Number(item.toString()));
};

export const getBigIntValue = (item: Codec): bigint => {
  return BigInt(getTextValue(item) || 0);
};

export const getBooleanValue = (item: Codec): boolean => {
  return JSON.parse(getTextValue(item));
};

export const hexToString = (input: string): string => {
  const hex = hexStripPrefix(input);
  let str = '';
  for (let i = 0; i < hex.length; i += 2)
    str += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
  return removeNullChars(str);
};

export const getSigner = (extrinsic: SubstrateExtrinsic): string => {
  return hexStripPrefix(
    u8aToHex(
      decodeAddress(
        extrinsic.extrinsic.signer.toString(),
        false,
        extrinsic.extrinsic.registry.chainSS58
      )
    )
  );
};

/**
 * Parses a raw Asset Document
 */
export const getDocValue = (
  doc: Codec
): Pick<AssetDocument, 'name' | 'link' | 'contentHash' | 'type' | 'filedAt'> => {
  const {
    uri: link,
    content_hash: documentHash,
    name,
    doc_type: type,
    filing_date: filedAt,
  } = JSON.parse(doc.toString());

  const hashType = Object.keys(documentHash)[0];

  return {
    name,
    link,
    contentHash: documentHash[hashType],
    type,
    filedAt,
  };
};

export const getSecurityIdentifiers = (item: Codec): SecurityIdentifier[] => {
  const identifiers = JSON.parse(item.toString());
  return identifiers.map(i => {
    const type = Object.keys(i)[0];
    return {
      type,
      value: i[type],
    };
  });
};

/**
 * Parses a Vec<AssetCompliance>
 */
export const getComplianceRulesValue = (
  requirements: Codec
): Pick<Compliance, 'complianceId' | 'data'>[] => {
  return JSON.parse(JSON.stringify(requirements));
};

/**
 * Parses AssetCompliance
 */
export const getComplianceValue = (
  compliance: Codec
): Pick<Compliance, 'complianceId' | 'data'> => {
  const { id, ...data } = JSON.parse(compliance.toString());
  return {
    complianceId: Number(id),
    data,
  };
};

/**
 * Parses AssetTransferManager
 */
export const getTransferManagerValue = (
  manager: Codec
): Pick<TransferManager, 'type' | 'value'> => {
  const { countTransferManager, percentageTransferManager } = JSON.parse(JSON.stringify(manager));

  if (countTransferManager) {
    return {
      type: TransferRestrictionTypeEnum.Count,
      value: Number(countTransferManager),
    };
  }

  if (percentageTransferManager) {
    return {
      type: TransferRestrictionTypeEnum.Percentage,
      value: Number(percentageTransferManager),
    };
  }

  throw new Error('Unknown transfer restriction type found');
};

export const getExemptKeyValue = (
  item: Codec
): Omit<Attributes<TransferComplianceExemption>, 'exemptedEntityId'> => {
  const {
    asset: { ticker },
    op: opType,
    claimType,
  } = JSON.parse(item.toString());
  return {
    assetId: hexToString(ticker),
    opType,
    claimType,
  };
};

export const getExemptionsValue = (exemptions: Codec): string[] => {
  return exemptions.toJSON() as string[];
};

export const logFoundType = (type: string, rawType: string): void => {
  FoundType.create({ id: type, rawType }).save();
};

export const END_OF_TIME = BigInt('253402194600000');

export function addIfNotIncludes<T>(arr: T[], item: T): void {
  if (!arr.includes(item)) {
    arr.push(item);
  }
}

interface MeshPortfolio {
  did: string;
  kind: {
    user: number;
  };
}

const meshPortfolioToPortfolio = ({
  did,
  kind: { user: number },
}: MeshPortfolio): Pick<Portfolio, 'identityId' | 'number'> => ({
  identityId: did,
  number: number || 0,
});

export const getPortfolioValue = (item: Codec): Pick<Portfolio, 'identityId' | 'number'> => {
  const meshPortfolio = JSON.parse(item.toString());
  return meshPortfolioToPortfolio(meshPortfolio);
};

export const getCaIdValue = (item: Codec): Pick<Distribution, 'localId' | 'assetId'> => {
  const { local_id: localId, ticker } = JSON.parse(item.toString());
  return { localId, assetId: hexToString(ticker) };
};

export interface LegDetails {
  from: Pick<Portfolio, 'identityId' | 'number'>;
  to: Pick<Portfolio, 'identityId' | 'number'>;
  ticker: string;
  amount: bigint;
}

export const getLegsValue = (item: Codec): LegDetails[] => {
  const legs = JSON.parse(item.toString());
  return legs.map(({ from: fromPortfolio, to: toPortfolio, asset: ticker, amount }) => ({
    from: meshPortfolioToPortfolio(fromPortfolio),
    to: meshPortfolioToPortfolio(toPortfolio),
    ticker: hexToString(ticker),
    amount: getBigIntValue(amount),
  }));
};

export const getSignerAddress = (event: SubstrateEvent): string => {
  let signer: string;
  if (event.extrinsic) {
    signer = getSigner(event.extrinsic);
  }
  return signer;
};

export const getDistributionValue = (
  item: Codec
): Pick<
  Distribution,
  'portfolioId' | 'currency' | 'perShare' | 'amount' | 'remaining' | 'paymentAt' | 'expiresAt'
> => {
  const { from, currency, per_share, amount, remaining, payment_at, expires_at } = JSON.parse(
    item.toString()
  );
  const { identityId, number } = meshPortfolioToPortfolio(from);
  return {
    portfolioId: `${identityId}/${number}`,
    currency: hexToString(currency),
    perShare: getBigIntValue(per_share),
    amount: getBigIntValue(amount),
    remaining: getBigIntValue(remaining),
    paymentAt: getBigIntValue(payment_at),
    expiresAt: getBigIntValue(expires_at || END_OF_TIME),
  };
};

export const logError = (message: string): void => {
  logger.error(message);
};
