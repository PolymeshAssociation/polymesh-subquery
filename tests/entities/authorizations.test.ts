import { gql } from '@apollo/client/core';
import { getApolloClient } from '../util';
const { query } = getApolloClient();

describe('authorizations', () => {
  it('should return all authorizations', async () => {
    const q = {
      query: gql`
        query {
          authorizations(first: 100) {
            nodes {
              authId: id
              type
              fromDid: fromId
              toDid: toId
              toKey
              data
              expiry
              status
              createdBlockId
              updatedBlockId
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();

    expect(subquery?.data).toMatchSnapshot();
  });
  it('should return authorisations filtered by authorization type', async () => {
    const q = {
      query: gql`
        query {
          authorizations(first: 100, filter: { type: { in: [BecomeAgent, AddMultiSigSigner] } }) {
            nodes {
              authId: id
              type
              fromDid: fromId
              toDid: toId
              toKey
              data
              expiry
              status
              createdBlockId
              updatedBlockId
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();

    expect(subquery?.data).toMatchSnapshot();
  });
});
