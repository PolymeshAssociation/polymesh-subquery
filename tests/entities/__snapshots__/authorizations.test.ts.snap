// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authorizations should return all authorizations 1`] = `
Object {
  "authorizations": Object {
    "__typename": "AuthorizationsConnection",
    "nodes": Array [
      Object {
        "__typename": "Authorization",
        "authId": "103",
        "createdBlockId": "128",
        "data": "{\\"asset\\":{\\"these\\":[]},\\"extrinsic\\":{\\"these\\":[]},\\"portfolio\\":{\\"these\\":[]}}",
        "expiry": null,
        "fromDid": "0xc5e2d554233da63d509ee482dbeed0ddec94dc1d0b45ebfdcdc48bd0928222b1",
        "status": "Consumed",
        "toDid": null,
        "toKey": "0xbc18d7b0241882bcfc94b5fe0d520988f2a9f4050a2a1df58c4de14ee23f3915",
        "type": "JoinIdentity",
        "updatedBlockId": "129",
      },
      Object {
        "__typename": "Authorization",
        "authId": "230",
        "createdBlockId": "311",
        "data": "{\\"0x31325449434b455200000000\\",\\"{\\\\\\"full\\\\\\":null}\\"}",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Consumed",
        "toDid": "0xeb686d5391c1123fcd252419bc1e77b525f46bddc8964d34e74be0988c419fea",
        "toKey": null,
        "type": "BecomeAgent",
        "updatedBlockId": "312",
      },
      Object {
        "__typename": "Authorization",
        "authId": "231",
        "createdBlockId": "313",
        "data": "{\\"0x31325449434b455200000000\\",\\"{\\\\\\"full\\\\\\":null}\\"}",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Consumed",
        "toDid": "0x0854afeca045161ef42ba9be94c973f65efc0e8532933500865a1fd655148f6c",
        "toKey": null,
        "type": "BecomeAgent",
        "updatedBlockId": "314",
      },
      Object {
        "__typename": "Authorization",
        "authId": "232",
        "createdBlockId": "317",
        "data": "{\\"0x31325449434b455200000000\\",\\"{\\\\\\"full\\\\\\":null}\\"}",
        "expiry": null,
        "fromDid": "0xeb686d5391c1123fcd252419bc1e77b525f46bddc8964d34e74be0988c419fea",
        "status": "Consumed",
        "toDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "toKey": null,
        "type": "BecomeAgent",
        "updatedBlockId": "318",
      },
      Object {
        "__typename": "Authorization",
        "authId": "247",
        "createdBlockId": "344",
        "data": "{\\"asset\\":{\\"these\\":[]},\\"extrinsic\\":{\\"these\\":[]},\\"portfolio\\":{\\"these\\":[]}}",
        "expiry": null,
        "fromDid": "0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e",
        "status": "Consumed",
        "toDid": null,
        "toKey": "0x6e041ae54374e2e593f540782a9c54d0a91ca8ebf91060acd5b69e7332e9856c",
        "type": "JoinIdentity",
        "updatedBlockId": "345",
      },
      Object {
        "__typename": "Authorization",
        "authId": "269",
        "createdBlockId": "722",
        "data": "{\\"asset\\":{\\"whole\\":null},\\"extrinsic\\":{\\"whole\\":null},\\"portfolio\\":{\\"whole\\":null}}",
        "expiry": null,
        "fromDid": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "status": "Consumed",
        "toDid": null,
        "toKey": "0xbc536d93907cfd70c5db6382d27776554f7338669c518dcef282caf7a4e8fc4b",
        "type": "JoinIdentity",
        "updatedBlockId": "723",
      },
      Object {
        "__typename": "Authorization",
        "authId": "277",
        "createdBlockId": "740",
        "data": "{\\"5Dfew5jnLuwQqEvvYFJcTJxHanfrNUGx33tWfbuR2a1X9oqe\\",\\"5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY\\",\\"100000\\"}",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Consumed",
        "toDid": null,
        "toKey": "0x46e3fad9460b2795400d9b6b93ac367e21dfcc22cb27d1adbcfd7a5187ed9d6f",
        "type": "AddRelayerPayingKey",
        "updatedBlockId": "741",
      },
      Object {
        "__typename": "Authorization",
        "authId": "75",
        "createdBlockId": "102",
        "data": "{\\"asset\\":{\\"these\\":[]},\\"extrinsic\\":{\\"these\\":[]},\\"portfolio\\":{\\"these\\":[]}}",
        "expiry": null,
        "fromDid": "0x25ebc60487a6e8c621b5987d0efa148a61be92f4810690b9a7d45298145f4ae8",
        "status": "Pending",
        "toDid": null,
        "toKey": "0x5a6a40ba68157d8c645bd9d380c0b1f638195740c6ab01510e66e2fb0fdb780e",
        "type": "JoinIdentity",
        "updatedBlockId": "102",
      },
      Object {
        "__typename": "Authorization",
        "authId": "76",
        "createdBlockId": "103",
        "data": "{\\"asset\\":{\\"these\\":[]},\\"extrinsic\\":{\\"these\\":[]},\\"portfolio\\":{\\"these\\":[]}}",
        "expiry": null,
        "fromDid": "0xda9b8b789c89450c4711db80f31b26c6d70b3075806a29806a9e0ab0dd72d88e",
        "status": "Pending",
        "toDid": null,
        "toKey": "0x7a37282c5bbeee50ef3f3fe0308d344284ac6384b1f75deb1caad9f86d095a08",
        "type": "JoinIdentity",
        "updatedBlockId": "103",
      },
      Object {
        "__typename": "Authorization",
        "authId": "77",
        "createdBlockId": "104",
        "data": "5DqZyPx9NAVzWhzRLeptzz9u1GwxsX5AQB3pzrUAHFFBJHCM",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Pending",
        "toDid": "0x139dad9675f6d89bbe39d4a1ffcb824b9ff44fce366390090752a926d7c6e689",
        "toKey": null,
        "type": "AddMultiSigSigner",
        "updatedBlockId": "104",
      },
      Object {
        "__typename": "Authorization",
        "authId": "78",
        "createdBlockId": "104",
        "data": "5DqZyPx9NAVzWhzRLeptzz9u1GwxsX5AQB3pzrUAHFFBJHCM",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Pending",
        "toDid": "0x81c9b01d440ae8b2330e400411564d64446502b1452cc31de731730e411a6e39",
        "toKey": null,
        "type": "AddMultiSigSigner",
        "updatedBlockId": "104",
      },
      Object {
        "__typename": "Authorization",
        "authId": "79",
        "createdBlockId": "104",
        "data": "5DqZyPx9NAVzWhzRLeptzz9u1GwxsX5AQB3pzrUAHFFBJHCM",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Pending",
        "toDid": "0xee7cd3d29832c3ad6db1dedb99b8fd146b57c6458fd79d17c5ac81f3d098227a",
        "toKey": null,
        "type": "AddMultiSigSigner",
        "updatedBlockId": "104",
      },
      Object {
        "__typename": "Authorization",
        "authId": "94",
        "createdBlockId": "116",
        "data": "{\\"asset\\":{\\"these\\":[]},\\"extrinsic\\":{\\"these\\":[]},\\"portfolio\\":{\\"these\\":[]}}",
        "expiry": null,
        "fromDid": "0xd066bfeefcb60c5a33bfa79d4466ad016a12e80fa0149b85f95884ed4327071e",
        "status": "Consumed",
        "toDid": null,
        "toKey": "0x460963cd0e0e6a8d4553ba2adde6cfa9e56786895f67f0a320967a44afd1e55c",
        "type": "JoinIdentity",
        "updatedBlockId": "118",
      },
      Object {
        "__typename": "Authorization",
        "authId": "95",
        "createdBlockId": "117",
        "data": "{\\"asset\\":{\\"these\\":[]},\\"extrinsic\\":{\\"these\\":[]},\\"portfolio\\":{\\"these\\":[]}}",
        "expiry": null,
        "fromDid": "0x6f6e1b8b1554b91f186a5040bad955e00b526e7d27dab20c0f4f4fdedd94cd0b",
        "status": "Consumed",
        "toDid": null,
        "toKey": "0xfaf1ab23de39dc22c7c56eb662c96e6e0ff527acffadc2d005ac3065f767cf78",
        "type": "JoinIdentity",
        "updatedBlockId": "119",
      },
    ],
  },
}
`;

exports[`authorizations should return authorisations filtered by authorization type 1`] = `
Object {
  "authorizations": Object {
    "__typename": "AuthorizationsConnection",
    "nodes": Array [
      Object {
        "__typename": "Authorization",
        "authId": "230",
        "createdBlockId": "311",
        "data": "{\\"0x31325449434b455200000000\\",\\"{\\\\\\"full\\\\\\":null}\\"}",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Consumed",
        "toDid": "0xeb686d5391c1123fcd252419bc1e77b525f46bddc8964d34e74be0988c419fea",
        "toKey": null,
        "type": "BecomeAgent",
        "updatedBlockId": "312",
      },
      Object {
        "__typename": "Authorization",
        "authId": "231",
        "createdBlockId": "313",
        "data": "{\\"0x31325449434b455200000000\\",\\"{\\\\\\"full\\\\\\":null}\\"}",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Consumed",
        "toDid": "0x0854afeca045161ef42ba9be94c973f65efc0e8532933500865a1fd655148f6c",
        "toKey": null,
        "type": "BecomeAgent",
        "updatedBlockId": "314",
      },
      Object {
        "__typename": "Authorization",
        "authId": "232",
        "createdBlockId": "317",
        "data": "{\\"0x31325449434b455200000000\\",\\"{\\\\\\"full\\\\\\":null}\\"}",
        "expiry": null,
        "fromDid": "0xeb686d5391c1123fcd252419bc1e77b525f46bddc8964d34e74be0988c419fea",
        "status": "Consumed",
        "toDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "toKey": null,
        "type": "BecomeAgent",
        "updatedBlockId": "318",
      },
      Object {
        "__typename": "Authorization",
        "authId": "77",
        "createdBlockId": "104",
        "data": "5DqZyPx9NAVzWhzRLeptzz9u1GwxsX5AQB3pzrUAHFFBJHCM",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Pending",
        "toDid": "0x139dad9675f6d89bbe39d4a1ffcb824b9ff44fce366390090752a926d7c6e689",
        "toKey": null,
        "type": "AddMultiSigSigner",
        "updatedBlockId": "104",
      },
      Object {
        "__typename": "Authorization",
        "authId": "78",
        "createdBlockId": "104",
        "data": "5DqZyPx9NAVzWhzRLeptzz9u1GwxsX5AQB3pzrUAHFFBJHCM",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Pending",
        "toDid": "0x81c9b01d440ae8b2330e400411564d64446502b1452cc31de731730e411a6e39",
        "toKey": null,
        "type": "AddMultiSigSigner",
        "updatedBlockId": "104",
      },
      Object {
        "__typename": "Authorization",
        "authId": "79",
        "createdBlockId": "104",
        "data": "5DqZyPx9NAVzWhzRLeptzz9u1GwxsX5AQB3pzrUAHFFBJHCM",
        "expiry": null,
        "fromDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "status": "Pending",
        "toDid": "0xee7cd3d29832c3ad6db1dedb99b8fd146b57c6458fd79d17c5ac81f3d098227a",
        "toKey": null,
        "type": "AddMultiSigSigner",
        "updatedBlockId": "104",
      },
    ],
  },
}
`;
